index.md

```diff
@@ -2,9 +2,286 @@
+
+We assume that there will be far fewer writes than there will be reads and most of the writes will involve multi-dataset transfers from personal workspaces or from user-batch outputs rather than people calling `butler.put()` from notebooks.
+
+There is still some debate over where user-generated products will be stored and that is discussed [later in this document](#writing-datasets).
+
+We have considered multiple options.
+
+#### Use primary database
+
+In this scenario, whilst queries can be spread across multiple database servers, writes always go to a single database server.
+Those updates are then replicated to the other servers, albeit with no immediate consistency guarantees and the possibility that someone could put a dataset and not be able to retrieve it straightaway due to replication latency.
+
+This is the simplest approach and closely matches how writes are handled in the direct butler.
+
+#### Use Butler Workspaces
+
+An option explored in {cite:p}`DMTN-271` is to use the concept of butler workspaces.
+In this scenario people use a local SQLite registry with their POSIX disk allocation, transferring dataset registry entries to their local butler, processing them locally, and keeping the output datasets local.
+This could be expanded to be a more permanent local storage solution, although there are limits as to how big a SQLite registry can usefully be.
+
+This approach does not work with the concept of allowing scientists to collaborate on their research since the datasets and registry would be completely private.
+It also does not work well with the concept of user batch {cite:p}`DMTN-223` where data are processed at the USDF but there is no way for a cloud RSP user to see the outputs if they are using a private Butler in the cloud.
+
+#### Butlers for Everyone
+
+An option proposed at the Joint Technical Meeting in February was to give everyone a personal butler PostgreSQL registry along with a personal object store bucket.[^oprah]
+This is intriguing because it allows for everyone to be completely independent, allows us to allocate users to specific servers, and would not require us to replicate all registry entries to every server.
+It would also be possible to create group-based butlers for people to collaborate on – if a scientist wants to make a dataset available to the group they can transfer the dataset to that butler using `Butler.transfer_from` in the normal way without changing the UUID of the dataset.
+
+Seeding these personal butlers would involve using `Butler.transfer_from` from the primary butler (but not copying any of the files, instead somehow enforcing a `direct` transfer mode be used), which leads to the possibility of someone seeding their butler with all of the datasets from the primary one.
+At that point you might have 10,000 or more personal replicas of the entire data release registry, which is less than ideal.
```

Comment:
My interpretation of the butlers-for-everyone concept was more along the lines of these being a more hosted version of the SQLite/POSIX repo concept, in which these personal data repositories would only have official data products that were explicitly transferred to them, with quotas prohibiting actually transferring everything.  The advantage over SQLite/POSIX repos would be that we could optimize transfers of database content and reduce actual copies of file content, and leave the door open to multi-registry queries a bit more open for the future.

For both the SQLite/POSIX concept and especially the hosted-butler-for-everyone concept, I had envisioned _also_ using the "primary database" or "single user registry" concept as a way to share user-created things between users, but that writes to the shared repository to only take the form of bulk transfers from user repositories, rather than direct write operations.  My hope was that limiting those writes would mitigate the problems associated with having a shared writeable repo to the point that we could manage them.

Reply:
I really thought at JTM we were discussing allocating a per-user schema in a cloudsql database for butlers-for-everyone, not local sqlite local posix storage.

I agree we should not allow transfers of files from primary to user butler but I was worried about registry entries for them. Hence the comment above about ending up with every user registry having all the primary records in it -- something we would have to consider blocking at the server (but which is much harder to block in a local sqlite but if it users their quota and makes things slow then we don't really care).

Reply:
> I really thought at JTM we were discussing allocating a per-user schema in a cloudsql database for butlers-for-everyone, not local sqlite local posix storage.

Yes, me too - that's what I meant by "more hosted version of" here - same kind of repo _content_ and relationship to the central repo as SQLite/POSIX, but backed by cloudsql database with object storage.

Reply:
What Jim said above roughly tracks with how I think this would work, if we go down this path.

Reply:
@TallJimbo I don't think you responded to my second comment.

> these personal data repositories would only have official data products that were explicitly transferred to them, with quotas prohibiting actually transferring everything.

I don't think quotas solve the problem because somebody could to a transfer_from using "direct" and not transfer any of the files at all but have all the registry records. That's what I was pondering with my "10,000 personal replicas" comment. I was suggesting we have a `transfer_from` that simply refuses to transfer any UUIDs that exist in the source butler. (which is what I explain in the line after this). We can't have hosted personal cloudsql registries if they all end up with copies of the original datastore registry records but it's fine if it only has things that people have generated themselves.

Reply:
I guess I was imagining that we'd have DB quotas, too, but I have no idea if that's actually feasible.  I suppose we could at least implement our own quotas on row counts of various types, which would be a decent proxy for actual resource-usage quotas, but it might be a lot of work.